### ISSUE TITLE
Found a bug when using reflecting BC in 2D RZ

### ISSUE THREAD

**Issue by Yin-YinjianZhao** [ECP-WarpX/WarpX#3821]
I found some particles that are supposed to be reflected with a specular reflection are not, their reflected speeds are somehow reduced such that a number of particles are accumulated and trapped on a very thin layer at the bigger r boundary.

As can be seen in this gif after about 6000 time steps.

![zoom1](https://user-images.githubusercontent.com/56095356/230698308-6fbbfeb6-9d2e-4a83-9b98-5fc6b229be64.gif)

As well as in this zoomed gif.

![zoom](https://user-images.githubusercontent.com/56095356/230698370-3901ec74-3649-4763-8271-7a51b7b0a1f3.gif)

I was using this picmy.py script:
[picmi.txt](https://github.com/ECP-WarpX/WarpX/files/11182317/picmi.txt)
in which electron's do_not_gather=1, such that they are moving without feeling the fields.

The running output reads:
[output.txt](https://github.com/ECP-WarpX/WarpX/files/11182320/output.txt)

I used the latest WarpX cloned yesterday.

Could someone help me take a look, it should be a small fix I believe.

@RemiLehe @dpgrote

**Comment by dpgrote** [comment]
I'll take a look at it.
    Dave

**Comment by Yin-YinjianZhao** [comment]
Then, I realized in 2D-RZ, waprx still pushes particles in 3D-XYZ coordinates, as commented in:
https://github.com/ECP-WarpX/WarpX/blob/development/Source/Particles/Pusher/UpdatePosition.H#L37

However, my old picmi script only gives initial x and z positions, I'm not sure if that is okay.
So, I tried to give x,y,z positions this time in this modified script.
[picmi.txt](https://github.com/ECP-WarpX/WarpX/files/11205988/picmi.txt)
In short,
```
>         x = rmin + r0*np.sqrt(np.random.rand(n_add))
>         y = np.zeros(n_add)
```
is replaced by
```
<         r = rmin + r0*np.sqrt(np.random.rand(n_add))
<         a = np.random.rand(n_add)*2*sc.pi
<         x = r * np.cos(a)
<         y = r * np.sin(a)
```

However, this time I got even worse resutls:
![zoom](https://user-images.githubusercontent.com/56095356/231318249-5ac8bf3d-9962-4b2b-b960-ec9a0d0599f0.gif)

But the good news is that at lest we are more clear that the bug is related to `y` in 2D-RZ.

@dpgrote What do you think?
BTW, where is the reflection BC written?
Here Source/Particles/ParticleBoundaries.cpp?

**Comment by Yin-YinjianZhao** [comment]
Well, then I tried to give only `y` and zero `x`.
```
<         r = rmin + r0*np.sqrt(np.random.rand(n_add))
<         a = np.random.rand(n_add)*2*sc.pi
<         x = r * 0.0
<         y = r * 1.0
```
And got all particles not being reflected at the boundary.
![zoom](https://user-images.githubusercontent.com/56095356/231321571-98b40370-64a8-40ca-a364-8b706e39be93.gif)
Therefore, it must be related to the reflection operation about `y` in 2D-RZ.

And because previously, I only give `x`, but after running for a while, some particles move along `y` as well, and those particles cause the trouble.

**Comment by Yin-YinjianZhao** [comment]
I noticed that, for example in `Source/Particles/ParticleBoundaries.cpp`, there are many `#ifdef WARPX_DIM_3D`,
but no `WARPX_DIM_RZ`, if `y` are updated in RZ as the same way in 3D, shouldn't we give something like
`#if (defined WARPX_DIM_3D || WARPX_DIM_RZ)`?
```
void
ParticleBoundaries::BuildReflectionModelParsers ()
{
    reflection_model_xlo_parser = std::make_unique<amrex::Parser>(
        utils::parser::makeParser(reflection_model_xlo_str, {"v"}));
    data.reflection_model_xlo = reflection_model_xlo_parser->compile<1>();
    reflection_model_xhi_parser = std::make_unique<amrex::Parser>(
        utils::parser::makeParser(reflection_model_xhi_str, {"v"}));
    data.reflection_model_xhi = reflection_model_xhi_parser->compile<1>();
#ifdef WARPX_DIM_3D
    reflection_model_ylo_parser = std::make_unique<amrex::Parser>(
        utils::parser::makeParser(reflection_model_ylo_str, {"v"}));
    data.reflection_model_ylo = reflection_model_ylo_parser->compile<1>();
    reflection_model_yhi_parser = std::make_unique<amrex::Parser>(
        utils::parser::makeParser(reflection_model_yhi_str, {"v"}));
    data.reflection_model_yhi = reflection_model_yhi_parser->compile<1>();
#endif
    reflection_model_zlo_parser = std::make_unique<amrex::Parser>(
        utils::parser::makeParser(reflection_model_zlo_str, {"v"}));
    data.reflection_model_zlo = reflection_model_zlo_parser->compile<1>();
    reflection_model_zhi_parser = std::make_unique<amrex::Parser>(
        utils::parser::makeParser(reflection_model_zhi_str, {"v"}));
    data.reflection_model_zhi = reflection_model_zhi_parser->compile<1>();
}
```

**Comment by Yin-YinjianZhao** [comment]
BTW, I know there is a 3D cylindrical pusher, which should be able to save some computational time when used in cylindrical PIC, and I once implemented it in one of my codes. Maybe it can contribute to WarpX as well.
https://www.sciencedirect.com/science/article/pii/S0021999113004798?via%3Dihub

**Comment by dpgrote** [comment]
I see what is happening. This is the result of having the velocities being in Cartesian coordinates when using RZ. In the boundary checking, with RZ, only the `x` position is checked, which is actually `r`. So when a particle reflects, the sign of only `ux` is changed. The code needs to be fixed so that the sign of `ur` is changed (and `ux` and `uy` updated accordingly). Note that there is the option `boundary.reflect_all_velocities` that will change the sign of all of the velocities, though that is not available yet in the Python interface. I'll submit a PR to fix this.

**Comment by Yin-YinjianZhao** [comment]
@dpgrote Great! Thanks!

BTW, if we still want to reflect both `x` and `y` accordingly, a specular reflection on a cylindrical surface may be needed.
I happen to implement this before in one of my codes.
Back then, I didn't find a good reference so derived it by myself, it should be correct, but requires some operations, I'm not sure if there is a smarter and faster method.

This file addresses the reflection on the bigger cylindrical `r` surface, which is what we needed.
A relevant derivation is in the appendix of a paper of mine (reflection on a spherical surface): https://doi.org/10.1017/s002237782100060x
[bc_reflect_dir2.txt](https://github.com/ECP-WarpX/WarpX/files/11216630/bc_reflect_dir2.txt)

This one addresses the reflection on the smaller cylindrical `r` surface.
[bc_reflect_dir1.txt](https://github.com/ECP-WarpX/WarpX/files/11216629/bc_reflect_dir1.txt)
And its derivation is in this png.
![reflection](https://user-images.githubusercontent.com/56095356/231603190-3af19b84-3db0-4430-a3e7-ee0382d72a0e.png)

**Comment by dpgrote** [comment]
Thanks for this derivation. I was going to use the existing code, which would do `r_reflected = 2*rmax - r_outside` and `theta_reflected = theta_outside`. This is simpler and also the previous location is not readily available. Because of the Courant condition, particles will only end up no more than one cell outside the domain so the error of using the approximation will be small.

**Comment by dpgrote** [comment]
This will be fixed in PR #3837
